src/Microsoft.Graph.Core/Requests/SimpleHttpProvider.cs

```diff
@@ -0,0 +1,200 @@
+        {
+            return this.SendAsync(request, HttpCompletionOption.ResponseContentRead, CancellationToken.None);
+        }
+
+        /// <summary>
+        /// Sends the request.
+        /// </summary>
+        /// <param name="request">The <see cref="HttpRequestMessage"/> to send.</param>
+        /// <param name="completionOption">The <see cref="HttpCompletionOption"/> to pass to the <see cref="IHttpProvider"/> on send.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> for the request.</param>
+        /// <returns>The <see cref="HttpResponseMessage"/>.</returns>
+        public async Task<HttpResponseMessage> SendAsync(
+            HttpRequestMessage request, 
+            HttpCompletionOption completionOption, 
+            CancellationToken cancellationToken)
+        {
+            var response = await this.SendRequestAsync(request, completionOption, cancellationToken).ConfigureAwait(false);
+
+            // check if the response is of a successful nature.
+            if (!response.IsSuccessStatusCode)
+            {
+                using (response)
+                {
+                    var errorResponse = await this.ConvertErrorResponseAsync(response).ConfigureAwait(false);
+                    Error error;
+
+                    if (errorResponse == null || errorResponse.Error == null)
+                    {
+                        if (response.StatusCode == HttpStatusCode.NotFound)
+                        {
+                            error = new Error { Code = ErrorConstants.Codes.ItemNotFound };
+                        }
+                        else
+                        {
+                            error = new Error
+                            {
+                                Code = ErrorConstants.Codes.GeneralException,
+                                Message = ErrorConstants.Messages.UnexpectedExceptionResponse,
+                            };
+                        }
+                    }
+                    else
+                    {
+                        error = errorResponse.Error;
+                    }
+
+                    if (string.IsNullOrEmpty(error.ThrowSite))
+                    {
+                        if (response.Headers.TryGetValues(CoreConstants.Headers.ThrowSiteHeaderName, out var throwSiteValues))
+                        {
+                            error.ThrowSite = throwSiteValues.FirstOrDefault();
+                        }
+                    }
+
+                    // Pass through the response headers and status code to the ServiceException.
+                    // System.Net.HttpStatusCode does not support RFC 6585, Additional HTTP Status Codes.
+                    // Throttling status code 429 is in RFC 6586. The status code 429 will be passed through.
+                    throw new ServiceException(error, response.Headers, response.StatusCode);
```

Comment:
Please review #37 as it adds some additional functionality to the HttpProvider that this provider should use to better support error/diagnostic info. Actually, I need a reviewer :). Let's pull 37 in and then we can pull yours in.

Reply:
That sounds like a plan.